### PR TITLE
refacor: Rewrite buffer cache

### PR DESCRIPTION
**Pull request by driftluo**
Considering that what we need is hierarchical ordering instead of global ordering, the original buffer strategy guarantees global ordering, and there will be additional consumption. For example, the buffers must be traversed all the time, and after changing the order to the hierarchy, this Side effects will disappear

dependence #243

### CODE REVIEW

File: src/session.rs [hunk 1]
```diff
@@ -870,8 +876,12 @@ where
...
             self.state,
             self.read_buf.len(),
-            self.write_buf.len(),
-            self.high_write_buf.len()
+            self.write_buf
+                .values()
+                .fold(0, |acc, item| acc + item.len()),
+            self.high_write_buf
+                .values()
+                .fold(0, |acc, item| acc + item.len())
```
Comment (quake) [inline src/session.rs:884]: same `log_enabled` issue here

File: yamux/src/session.rs [hunk 2]
```diff
@@ -365,81 +365,86 @@ where
...
+            }
+            if frames.capacity() > frames.len() + BUF_SHRINK_THRESHOLD {
+                frames.shrink_to_fit();
             }
         }
-
-        if self.read_pending_frames.capacity() > BUF_SHRINK_THRESHOLD {
-            self.read_pending_frames.shrink_to_fit();
+        for id in disconnect_streams {
+            self.read_pending_frames.remove(&id);
```
Comment (quake) [inline yamux/src/session.rs:435]: should be removing from buf?
```
  buf.remove(&id);
```

Comment (driftluo) [reply yamux/src/session.rs:435]: Yes, this is a mistake

File: src/service.rs [hunk 3]
```diff
@@ -359,64 +359,70 @@ where
...
-        if priority.is_high() {
-            self.high_write_buf.push_back((id, event));
-        } else {
-            self.write_buf.push_back((id, event));
-        }
-    }
-
     #[inline(always)]
-    fn distribute_to_session_process<D: Iterator<Item = (SessionId, SessionEvent)>>(
+    fn distribute_to_session_process(
```
Comment (doitian) [inline src/service.rs:363]: Not related to this PR: I think we should limit the number of events to dispatch in one round.

Comment (driftluo) [reply src/service.rs:363]: In fact, here's what I should say: under low stress, whatever you do is not a problem, but under prolonged high stress, whatever you do is wrong

**Pull request merged**